pool/paymentmgr.go

```diff
@@ -595,71 +644,237 @@ func (pm *PaymentMgr) maturePendingPayments(height uint32) (map[uint32][]*Paymen
-		if updatedFee == 0 {
-			delete(pmtDetails, addr.String())
-		} else {
-			pmtDetails[addr.String()] = updatedFee
+
+	for _, set := range pmts {
+		index := uint32(2)
+		txHash, err := chainhash.NewHashFromStr(set[0].Source.Coinbase)
+		if err != nil {
+			return err
 		}
-	}
-	pmts := make(map[dcrutil.Address]dcrutil.Amount, len(pmtDetails))
-	for dest, amt := range pmtDetails {
-		addr, err := dcrutil.DecodeAddress(dest, pm.cfg.ActiveNet)
+
+		txOutResult, err := txCreator.GetTxOut(txHash, index, false)
+		if err != nil {
+			return fmt.Errorf("unable to find tx output: %v", err)
+		}
+
+		// Ensure the referenced prevout to be spent is a coinbase and
+		// spendable at the current height.
+		if !txOutResult.Coinbase {
+			return fmt.Errorf("expected the referenced output at index %d "+
+				"for tx %v to be a coinbase", index, txHash.String())
+		}
+
+		if txOutResult.Confirmations < int64(pm.cfg.ActiveNet.CoinbaseMaturity+1) {
+			return fmt.Errorf("expected the referenced coinbase at index %d "+
+				"for tx %v to be spendable", index, txHash.String())
+		}
+
+		in := chainjson.TransactionInput{
+			Amount: txOutResult.Value,
+			Txid:   txHash.String(),
+			Vout:   index,
+			Tree:   wire.TxTreeRegular,
+		}
+		inputs = append(inputs, in)
+
+		outV, err := dcrutil.NewAmount(txOutResult.Value)
 		if err != nil {
 			return err
 		}
-		pmts[addr] = amt
+		tIn += outV
+
+		// Generate the outputs paying dividends and fees.
+		for _, pmt := range set {
+			if pmt.Account == poolFeesK {
+				_, ok := outputs[feeAddr.String()]
+				if !ok {
+					outputs[feeAddr.String()] = pmt.Amount
+					tOut += pmt.Amount
+					continue
+				}
+				outputs[feeAddr.String()] += pmt.Amount
+				tOut += pmt.Amount
+				continue
+			}
+
+			acc, err := FetchAccount(pm.cfg.DB, []byte(pmt.Account))
+			if err != nil {
+				return err
+			}
+
+			_, ok := outputs[acc.Address]
+			if !ok {
+				outputs[acc.Address] = pmt.Amount
+				tOut += pmt.Amount
+				continue
+			}
+			outputs[acc.Address] += pmt.Amount
+			tOut += pmt.Amount
+		}
+	}
+
+	var diff dcrutil.Amount
+	if tOut > tIn {
+		diff = tOut - tIn
+	} else {
+		diff = tIn - tOut
+	}
+
+	// Ensure the generated outputs are equal to the inputs being spent.
+	if diff != zeroAmount && diff > maxValueDifference {
+		return fmt.Errorf("total input %s and output %s value mismatch",
+			tIn, tOut)
+	}
+
+	inSizes := make([]int, len(inputs))
+	for range inputs {
+		inSizes = append(inSizes, txsizes.RedeemP2PKHSigScriptSize)
+	}
+
+	outSizes := make([]int, len(outputs))
+	for range outputs {
+		inSizes = append(inSizes, txsizes.P2PKHOutputSize)
 	}
 
-	txid, err := pm.cfg.PublishTransaction(pmts, *targetAmt)
+	estSize := txsizes.EstimateSerializeSizeFromScriptSizes(inSizes, outSizes,
+		txsizes.P2PKHOutputSize)
+	estFee := txrules.FeeForSerializeSize(txrules.DefaultRelayFeePerKb, estSize)
+
+	// Deduct the transaction fees from the pool fees being paid out.
+	fees := outputs[feeAddr.String()]
+	fees = fees - estFee
```

Comment:
Do we need to handle the case where the pool fee is not big enough to cover the transaction fee? 

It is possible to set pool fee to any value over zero, for example 0.0000001%

Reply:
Good catch, will have to think about that some more. What's coming to mind immediately is to split the fees based on the percentages of the amounts being paid out compared to the coinbase total minus the pool fees.